autoload/lsp/ui/vim/signs.vim

```diff
@@ -88,60 +92,62 @@ function! lsp#ui#vim#signs#previous_error() abort
         return
     endif
 
-    let l:view.lnum = l:next_line
-    let l:view.topline = 1
+    let l:view['lnum'] = l:next_line
+    let l:view['topline'] = 1
     let l:height = winheight(0)
     let totalnum = line('$')
     if totalnum > l:height
         let l:half = l:height / 2
-        if l:totalnum - l:half < l:view.lnum
-            let l:view.topline = l:totalnum - l:height + 1
+        if l:totalnum - l:half < l:view['lnum']
+            let l:view['topline'] = l:totalnum - l:height + 1
         else
-            let l:view.topline = l:view.lnum - l:half
+            let l:view['topline'] = l:view['lnum'] - l:half
         endif
     endif
     call winrestview(l:view)
 endfunction
 
 " Set default sign text to handle case when user provides empty dict
 function! s:add_sign(sign_name, sign_default_text, sign_options) abort
-    let l:sign_string = 'sign define ' . a:sign_name
-    let l:sign_string .= ' text=' . get(a:sign_options, 'text', a:sign_default_text)
+    let l:options = {
+        \ 'text': get(a:sign_options, 'text', a:sign_default_text),
+        \ 'texthl': a:sign_name . 'Text',
+        \ 'linehl': a:sign_name . 'Line',
+        \ }
     let l:sign_icon = get(a:sign_options, 'icon', '')
     if !empty(l:sign_icon)
-        let l:sign_string .= ' icon=' . l:sign_icon
+        let l:options['icon'] = l:sign_icon
     endif
-    let l:sign_string .= ' texthl=' . a:sign_name . 'Text'
-    let l:sign_string .= ' linehl=' . a:sign_name . 'Line'
-    exec l:sign_string
+    call sign_define(a:sign_name, l:options)
 endfunction
 
 function! s:define_signs() abort
-    if !s:signs_defined
-        call s:add_sign('LspError', 'E>', g:lsp_signs_error)
-        call s:add_sign('LspWarning', 'W>', g:lsp_signs_warning)
-        call s:add_sign('LspInformation', 'I>', g:lsp_signs_information)
-        call s:add_sign('LspHint', 'H>', g:lsp_signs_hint)
-        let s:signs_defined = 1
-    endif
+    " let vim handle errors/duplicate instead of us maintaining the state
+    call s:add_sign('LspError', 'E>', g:lsp_signs_error)
+    call s:add_sign('LspWarning', 'W>', g:lsp_signs_warning)
+    call s:add_sign('LspInformation', 'I>', g:lsp_signs_information)
+    call s:add_sign('LspHint', 'H>', g:lsp_signs_hint)
+endfunction
+
+function! s:get_signs(bufnr) abort
+    let l:signs = sign_getplaced(a:bufnr, { 'group': '*' })
```

Comment:
please add below since some of users have mappings for next_error/previous_error in vimrc.

```
    if !s:supports_signs
        return []
    endif
```

Reply:
added check everywhere.